charmcraft/commands/store/registry.py

```diff
@@ -440,3 +448,77 @@ def _upload_blob(self, filepath, size, digest):
 
         # finally remove the temp filepath
         os.unlink(filepath)
+
+    def upload_from_local(self, digest: str) -> str:
+        """Upload the image from the local registry."""
+        dockerd = LocalDockerdInterface()
+
+        # validate the image is present locally
+        logger.debug("Checking image is present locally")
+        image_info = dockerd.get_image_info(digest)
+        if image_info is None:
+            return
+        local_image_size = image_info["Size"]
+
+        logger.debug("Getting the image from the local repo; size=%s", local_image_size)
+        response = dockerd.get_streamed_image_content(digest)
+
+        tmp_exported = tempfile.NamedTemporaryFile(mode="wb")
+        extracted_total = 0
+        for chunk in response.iter_content(2 ** 20):
+            extracted_total += len(chunk)
+            progress = 100 * extracted_total / local_image_size
+            print("Extracting... {:.2f}%\r".format(progress), end="", flush=True)
+            tmp_exported.file.write(chunk)
+
+        # open the image tar and inspect it to get the config and layers for the manifest
+        tmp_exported.file.flush()
+        image_tar = tarfile.open(tmp_exported.name)
+        tmp_exported.close()  # closing implies deletion, but the tar lib already grabbed it ok
```

Comment:
Will this work on Windows?  IIRC it doesn't allow open files to be deleted.

Reply:
Furthermore, it does not allow an open file to be opened! So it's not that I just can close it later. 

So, I had to tell the temp file to not delete on close, and handle that manually later.

The equivalent to this code (which is the one I wrote to try it on Windows):

```
import os
import tempfile

tmp_exported = tempfile.NamedTemporaryFile(mode="wb", delete=False)
tmp_exported.file.write(b"1234567\n")
tmp_exported.close()

tmp_path = tmp_exported.name

image_tar = open(tmp_path, 'rb')

print("Content:", image_tar.read())
print("Deleted?", not os.path.exists(tmp_path))
```

Anyway, fixed.